crypto/ts/quinTree.ts

```diff
@@ -0,0 +1,377 @@
+import type { Leaf, Node, IMerkleProof } from "./types";
+
+/**
+ * Calculate the zeroes and the root of a tree
+ * @param arity The arity of the tree
+ * @param depth The depth of the tree
+ * @param zeroValue The zero value of the tree
+ * @param hashFunc The hash function of the tree
+ * @returns The zeros and the root
+ */
+const calcInitialVals = (
+  arity: number,
+  depth: number,
+  zeroValue: bigint,
+  hashFunc: (leaves: bigint[]) => bigint,
+): { zeros: bigint[]; root: bigint } => {
+  const zeros: bigint[] = [];
+  let currentLevelHash = zeroValue;
+
+  for (let i = 0; i < depth; i += 1) {
+    zeros.push(currentLevelHash);
+
+    const z = Array<bigint>(arity).fill(currentLevelHash);
+
+    currentLevelHash = hashFunc(z);
+  }
+
+  return { zeros, root: currentLevelHash };
+};
+
+/**
+ * An implementation of an incremental Merkle tree
+ * @dev adapted from https://github.com/weijiekoh/optimisedmt
+ */
+export class MACITree {
+  // how many levels
+  depth: number;
+
+  // the zero value
+  zeroValue: bigint;
+
+  // the number of leaves per node
+  arity: number;
+
+  // the hash function used in the tree
+  hashFunc: (leaves: Leaf[]) => bigint;
+
+  // The the smallest empty leaf index
+  nextIndex = 0;
+
+  // Contains the zero value per level. i.e. zeros[0] is zeroValue,
+  // zeros[1] is the hash of leavesPerNode zeros, and so on.
+  zeros: bigint[] = [];
+
+  root: bigint;
+
+  nodes: Node;
+
+  numNodes: number;
+
+  capacity: number;
+
+  /**
+   * Create a new instance of the MaciQuinTree
+   * @param depth The depth of the tree
+   * @param zeroValue The zero value of the tree
+   * @param arity The arity of the tree
+   * @param hashFunc The hash function of the tree
+   */
+  constructor(depth: number, zeroValue: bigint, arity: number, hashFunc: (leaves: bigint[]) => bigint) {
+    this.depth = depth;
+    this.zeroValue = zeroValue;
+    this.arity = arity;
+    this.hashFunc = hashFunc;
+
+    // calculate the initial values
+    const { zeros, root } = calcInitialVals(this.arity, this.depth, this.zeroValue, this.hashFunc);
+    this.zeros = zeros;
+    this.root = root;
+
+    // calculate the number of nodes
+    let numNodes = 1;
+    for (let i = 0; i < this.depth; i += 1) {
+      numNodes += this.arity ** (i + 1);
+    }
+    this.numNodes = numNodes;
```

Comment:
I think this can be optimized to just one formula.
```suggestion
    this.numNodes = (this.arity ** (this.depth + 1)) / (this.arity - 1);
```